unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,723 @@
+        blocks_received.extend(last_blocks);
+
+        blocks_received
+    }
+}
+
+fn render_and_hash(flat: &FlatUnixFs<'_>) -> (Cid, Vec<u8>) {
+    // TODO: as shown in later dagger we don't really need to render the FlatUnixFs fully; we could
+    // either just render a fixed header and continue with the body OR links, though the links are
+    // a bit more complicated.
+    let mut out = Vec::with_capacity(flat.get_size());
+    let mut writer = Writer::new(&mut out);
+    flat.write_message(&mut writer)
+        .expect("unsure how this could fail");
+    let mh = multihash::wrap(multihash::Code::Sha2_256, &Sha256::digest(&out));
+    let cid = Cid::new_v0(mh).expect("sha2_256 is the correct multihash for cidv0");
+    (cid, out)
+}
+
+/// Chunker strategy
+#[derive(Debug, Clone)]
+pub enum Chunker {
+    /// Size based chunking
+    Size(usize),
+}
+
+impl std::default::Default for Chunker {
+    /// Returns a default chunker which matches go-ipfs 0.6
+    fn default() -> Self {
+        Chunker::Size(256 * 1024)
+    }
+}
+
+impl Chunker {
+    fn accept<'a>(&mut self, input: &'a [u8], buffered: &[u8]) -> (&'a [u8], bool) {
+        use Chunker::*;
+
+        match self {
+            Size(max) => {
+                let l = input.len().min(*max - buffered.len());
+                let accepted = &input[..l];
+                let ready = buffered.len() + l >= *max;
+                (accepted, ready)
+            }
+        }
+    }
+
+    fn size_hint(&self) -> usize {
+        use Chunker::*;
+
+        match self {
+            Size(max) => *max,
+        }
+    }
+}
+
+/// Collector or layout strategy
+#[derive(Debug, Clone)]
+pub enum Collector {
+    /// Balanced trees.
+    Balanced(BalancedCollector),
+}
+
+impl std::default::Default for Collector {
+    fn default() -> Self {
+        Collector::Balanced(Default::default())
+    }
+}
+
+impl Collector {
+    fn flush_links(&mut self, pending: &mut Vec<Link>, finishing: bool) -> Vec<(Cid, Vec<u8>)> {
+        use Collector::*;
+
+        match self {
+            Balanced(bc) => bc.flush_links(pending, finishing),
+        }
+    }
+}
+
+/// BalancedCollector creates balanced UnixFs trees, most optimized for random access to different
+/// parts of the file. Currently supports only link count threshold or the branching factor.
+#[derive(Clone)]
+pub struct BalancedCollector {
+    branching_factor: usize,
+    // reused between link block generation
+    reused_links: Vec<PBLink<'static>>,
+    // reused between link block generation
+    reused_blocksizes: Vec<u64>,
+}
+
+impl fmt::Debug for BalancedCollector {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "BalancedCollector {{ branching_factor: {} }}",
+            self.branching_factor
+        )
+    }
+}
+
+impl std::default::Default for BalancedCollector {
+    fn default() -> Self {
+        Self::with_branching_factor(174)
```

Comment:
what's the rationale behind that number? as usual, I'm a fan of well-visible `const`s for those, but a comment would work too

Reply:
I thought I documented somewhere but perhaps I only did it for chunker::default. I'll add a bit more.

Reply:
Does this look better? Added stuff for the enum Collector and here?

Reply:
yep, looking good :+1:

Reply:
> what's the rationale behind that number?

Obligatory: this number is only relevant to the current defaults of IPFS. With the power of hindsight it is definitely the wrong choice: do not hardcode it too deep, as you will end up changing it within a year ( plus the balanced strategy may get a makeover )